ee/tools/puppet/fleetdm/lib/puppet/functions/fleetdm/preassign_profile.rb

```diff
@@ -12,14 +12,21 @@
   end
 
   def preassign_profile(profile_identifier, host_uuid, template, group = 'default', ensure_profile = 'present')
-    host = call_function('lookup', 'fleetdm::host')
-    token = call_function('lookup', 'fleetdm::token')
-    client = Puppet::Util::FleetClient.new(host, token)
+    client = Puppet::Util::FleetClient.instance
     run_identifier = "#{closure_scope.catalog.catalog_uuid}-#{Puppet[:node_name_value]}"
     response = client.preassign_profile(run_identifier, host_uuid, template, group, ensure_profile)
 
     if response['error'].empty?
-      Puppet.info("successfully pre-assigned profile #{profile_identifier}")
+      base64_checksum = Digest::MD5.base64digest(template)
+      host = client.get_host_by_identifier(host_uuid)
+      host_profiles = client.get_host_profiles(host['body']['host']['id'])
+
+      if host_profiles['error'].empty?
+        unless host_profiles['body']['profiles'].any? { |p| p['checksum'] == base64_checksum }
```

Comment:
I might be missing something, but should the unless condition be all rather than any?

Reply:
we only want to execute this branch if the profile being preassigned doesn't match an existing profile, so we want to bail if any of the items match. Does that make sense?

I think I'm going to add a comment. We might even want to log regardless

Reply:
As I read it currently, we set `changed = true` only if none of the host profiles match the checksum. Is that right?

Reply:
that's right

Reply:
So if one host profile matches and one doesn't match, we would consider that host unchanged?

Reply:
It would be considered as changed, keep in mind that this function runs once per assigned profile

Reply:
Sorry if I'm being dense, but let's say `host_profiles['body']['profiles']` has two elements: the first matches the checksum and the second does not. So we would considered that host changed.

But seems like the `unless` causes cause the function to bail out before the `changed` flag is flipped to `true`. Are you saying that is ok?

Reply:
You're not being dense, don't worry!

One thing that I think it's confusing you: setting `changed = true` here marks the _resource_ as changed (which also means that the host has changed, but it's not the same thing)

This function is called internally by the `fleetdm::profile` resource, each time it's declared. So for example in the following manifest:

```pp
fleetdm::profile { 'my.ident.1': }
fleetdm::profile { 'my.ident.2': }
```

The code you're looking at will be called twice, one for each profile.

---

Now let's say that `host_profiles['body']['profiles']` returns an array with:

```json
[ { "identifier": "my.ident.1" } ]
```

1. When `fleetdm::profile { 'my.ident.1': }` is evaluated, this will bail out early, we don't want to mark the profile as changed because it was already there.
2. When `fleetdm::profile { 'my.ident.2': }` is evaluated, the branch is evaluated and we set that profile as changed

---

I think one thing we could improve is to tweak the language so it doesn't imply that a profile that changed is a profile that's added (because it might have `ensure => absent`)

Reply:
Thanks, that all makes sense! Just wanted to make sure since the booleans were tricky and the names were blurring together without all the extra context.

Reply:
sounds good, I sent a commit with some changes that'll hopefully help with that.